distributed/scheduler.py

```diff
@@ -4733,70 +4770,33 @@ def stimulus_task_erred(
-
         ts: TaskState = parent._tasks.get(key)
-        if ts is None:
-            return recommendations, client_msgs, worker_msgs
-
-        if ts._state == "processing":
-            retries: Py_ssize_t = ts._retries
-            r: tuple
-            if retries > 0:
-                ts._retries = retries - 1
-                r = parent._transition(key, "waiting")
-            else:
-                r = parent._transition(
-                    key,
-                    "erred",
-                    cause=key,
-                    exception=exception,
-                    traceback=traceback,
-                    worker=worker,
-                    **kwargs,
-                )
-            recommendations, client_msgs, worker_msgs = r
+        if ts is None or ts._state != "processing":
+            return {}, {}, {}
 
-        return recommendations, client_msgs, worker_msgs
+        if ts._retries > 0:
+            ts._retries -= 1
+            return parent._transition(key, "waiting")
+        else:
+            return parent._transition(
+                key,
+                "erred",
+                cause=key,
+                exception=exception,
+                traceback=traceback,
+                worker=worker,
+                **kwargs,
+            )
 
-    def stimulus_missing_data(
-        self, cause=None, key=None, worker=None, ensure=True, **kwargs
-    ):
-        """Mark that certain keys have gone missing.  Recover."""
+    def stimulus_missing_data(self, key=None):
+        """Mark that certain keys have gone missing. Recover."""
         parent: SchedulerState = cast(SchedulerState, self)
         with log_errors():
-            logger.debug("Stimulus missing data %s, %s", key, worker)
-
-            recommendations: dict = {}
-            client_msgs: dict = {}
-            worker_msgs: dict = {}
-
+            logger.debug("Stimulus missing data: %r", key)
             ts: TaskState = parent._tasks.get(key)
             if ts is None or ts._state == "memory":
-                return recommendations, client_msgs, worker_msgs
-            cts: TaskState = parent._tasks.get(cause)
-
-            if cts is not None and cts._state == "memory":  # couldn't find this
-                ws: WorkerState
-                cts_nbytes: Py_ssize_t = cts.get_nbytes()
-                for ws in cts._who_has:  # TODO: this behavior is extreme
-                    del ws._has_what[ts]
-                    ws._nbytes -= cts_nbytes
-                cts._who_has.clear()
-                recommendations[cause] = "released"
-
-            if key:
-                recommendations[key] = "released"
-
-            parent._transitions(recommendations, client_msgs, worker_msgs)
-            recommendations = {}
-
-            if parent._validate:
-                assert cause not in self.who_has
-
-            return recommendations, client_msgs, worker_msgs
+                return {}, {}, {}
+            return {key: "released"}, {}, {}
```

Comment:
I found myself more than once in a situation where I wanted to delete this stimulus handler since I believe most of it is dead code. However, I don't see the connection to the current PR. Is this somehow related? Can this change be split off?

Reply:
Actually I can't. If you read the old code, it does something clearly wrong - on lines 4783:4787, it deletes ts from has_what but it removes ``cts.nbytes`` from ``ws.nbytes``; it should have removed ``ts.nbytes`` instead. With nothing invoking this code path, I had no idea what the original intent of the logic was, so I opted for cutting the method down to just what is actually invoked by ``handle_release_data``.

Reply:
if you prefer I can remove the ``stimulus_missing_data`` method altogether and inline the surviving code into ``handle_release_data``?

Reply:
Maybe that's reasonable, since it's not really a `stimulus` function anymore either and it's only a few lines.

Reply:
I'm fine with moving the code to `handle_release_data`

Reply:
done; please double check that the new code is functionally identical to the previous one